### PR TITLE
Added Alibaba Cloud CS Content

### PR DESCRIPTION
**Pull request by sushilMF**
Created CS operations for the following:

Create Instance
Delete Instance
Start Instance
Stop Instance
Restart Instance
Allocate IP Address
Get Instance Status
Check Instance Status
Also Created Workflows for the following:

Deploy Instance
Undeploy Instance
Start Instance
Stop Instance
Reboot Instance

### CODE REVIEW

File: content/io/cloudslang/alibaba/ecs/start_instance.sl [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    - instance_id
+    - init_local_disk:
+        required: false
+    - polling_interval:
+        required: false
+    - polling_retries:
+        required: false
+
+  workflow:
+    - get_instance_Status:
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/start_instance.sl]: Please maintain lowercase for steps: change get_instance_Status to get_instance_status in all the places where this applies

File: content/io/cloudslang/alibaba/ecs/instances/check_instance_state.sl [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+          utils.sleep:
+            - seconds: ${get('polling_interval', '10')}
+        navigate:
+          - SUCCESS: FAILURE
+          - FAILURE: on_failure
+
+  outputs:
+    - output: ${return_result}
+    - return_code
+    - exception
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/check_instance_state.sl]: A good suggestion would be to have the instance_status output exposed as well.

File: content/io/cloudslang/alibaba/ecs/stop_instance.sl [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        publish:
+          - output: 'Instance is already in Stopped state'
+          - return_code
+          - exception
+          - instance_status_returned: '${instance_status}'
+        navigate:
+          - SUCCESS: string_occurrence_counter
+          - FAILURE: on_failure
+
+    - string_occurrence_counter:
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/stop_instance.sl]: Please change the step: string_occurrence_counter to a more meaningful step name, in all the flows where this is present.

File: content/io/cloudslang/alibaba/ecs/stop_instance.sl [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            - output
+            - return_code
+            - exception
+        navigate:
+          - SUCCESS: SUCCESS
+          - FAILURE: on_failure
+
+  outputs:
+    - return_result: '${output}'
+    - return_code
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/stop_instance.sl]: Please return the instance status as an output

File: content/io/cloudslang/alibaba/ecs/undeploy_instance.sl [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        publish:
+          - return_result
+          - return_code
+          - exception
+          - instance_status: '${instance_status}'
+        navigate:
+          - SUCCESS: string_occurrence_counter
+          - FAILURE: on_failure
+
+    - string_occurrence_counter:
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/undeploy_instance.sl]: Same here

File: content/io/cloudslang/alibaba/ecs/undeploy_instance.sl [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+#!
+#! @output return_result: The authentication token in case of success, or an error message in case of failure.
+#! @output return_code: "0" if operation was successfully executed, "-1" otherwise
+#! @output exception: Exception if there was an error when executing, empty otherwise
+#!
+#! @result SUCCESS: The server (instance) has been successfully terminated
+#! @result FAILURE: An error occur while trying to delete the instance
+#!!#
+########################################################################################################################
+
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/undeploy_instance.sl]: Although it stops the instance and then deletes it, an extra helpful step would be to check that the instance does not exist anymore.

File: content/io/cloudslang/alibaba/ecs/undeploy_instance.sl [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+#!                          Default: 10
+#! @input polling_retries: The number of retries to check if the instance is stopped.
+#!                         Default: 50
+#!
+#! @output return_result: The authentication token in case of success, or an error message in case of failure.
+#! @output return_code: "0" if operation was successfully executed, "-1" otherwise
+#! @output exception: Exception if there was an error when executing, empty otherwise
+#!
+#! @result SUCCESS: The server (instance) has been successfully terminated
+#! @result FAILURE: An error occur while trying to delete the instance
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/undeploy_instance.sl]: occurred

File: content/io/cloudslang/alibaba/ecs/reboot_instance.sl [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            - output
+            - return_code
+            - exception
+        navigate:
+          - SUCCESS: SUCCESS
+          - FAILURE: on_failure
+
+  outputs:
+    - output
+    - return_code
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/reboot_instance.sl]: Same here, publish as an output the instance state.

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+        required: false  
+        sensitive: true
+    - password_inherit:
+        default: "false"
+        required: false  
+    - passwordInherit: 
+        default: ${get('password_inherit', '')}  
+        required: false 
+        private: true 
+    - is_optimized:
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: is_optimized is missing default value

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+#!                  lowercase letters, digits, and special characters.Special characters such as ( ) ' ~ !  @ # $ % ^ &
+#!                  * - + = | { } [ ] : ; ‘ < > , . ? are allowed. /
+#!                  Optional
+#! @input password_inherit: Whether to use the password pre-configured in the image you select or not. When
+#!                          PasswordInherit is specified, the Password must be null. For a secure access, make sure that
+#!                          the selected image has password configured.
+#!                          Default: false
+#!                          Optional
+#! @input is_optimized: Whether it is an I/O-optimized instance or not. For phased-out instance types, the default value
+#!                      is none. For other instance types, the default value is optimized.
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Remove: ' the default value is optimized.' and add 

Default: 'optimized'

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl (hunk 9 above)
Comment (sushilMF) [reply content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Default value for is_optimized is not there. As per description i understood that there is default values is for phased-out instance and other instances types. So there is no default value only valid values are there.

**Comment by tethryus** [comment]
Hello, please update the cs-alibaba reference to the newly released version 0.0.2

**Comment by sushilMF** [comment]
Hello @tethryus , We have updated the cs-alibaba reference in all operations. Please review and merge.

File: content/io/cloudslang/alibaba/ecs/deploy_instance.sl [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+#! @input client_token: It is used to guarantee the idempotence of the request. This parameter value is generated by the
+#!                      client and is guaranteed to be unique between different requests. It can contain a maximum of 64
+#!                      ASCII characters only.
+#!                      Optional
+#! @input key_pair_name: The name of the key pair.This parameter is valid only for a Linux instance. For a Windows ECS
+#!                      instance, if a value is set for parameter KeyPairName, the password still takes effect. If a
+#!                      value is set for parameter KeyPairName, the Password still takes effect.The user name and
+#!                      password authentication method is disabled if a value is set for parameter KeyPairName for a
+#!                      Linux instance. 
+#!						Default: ''.
```
Comment (ionutmarisca) [inline content/io/cloudslang/alibaba/ecs/deploy_instance.sl]: Small indentation issue

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+#! @input client_token: It is used to guarantee the idempotence of the request. This parameter value is generated by the
+#!                      client and is guaranteed to be unique between different requests. It can contain a maximum of 64
+#!                      ASCII characters only. 
+#!                      Optional
+#! @input key_pair_name: The name of the key pair.This parameter is valid only for a Linux instance. For a Windows ECS
+#!                      instance, if a value is set for parameter KeyPairName, the password still takes effect. If a
+#!                      value is set for parameter KeyPairName, the Password still takes effect.The user name and
+#!                      password authentication method is disabled if a value is set for parameter KeyPairName for a
+#!                      Linux instance. 
+#!						Default: ''.
```
Comment (ionutmarisca) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Small indentation issue

File: content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl [hunk 13]
```diff
@@ -0,0 +1,127 @@
+########################################################################################################################
+#!!
+#! @description: This operation is used to stop an ECS instance.
+#!
+#! @input access_key_id: The Access Key ID associated with your Alibaba cloud account.
+#! @input access_key_secret: The Secret ID of the Access Key associated with your Alibaba cloud account.
+#! @input proxy_host: Proxy server used to access the Alibaba cloud services.
+#!                    Optional
+#! @input proxy_port: Proxy server port used to access the Alibaba cloud services.Default: '8080'
```
Comment (ionutmarisca) [inline content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl]: The default here should be on a new line

File: content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+#!                    Optional
+#! @input proxy_username: Proxy server user name.
+#!                        Optional
+#! @input proxy_password: Proxy server password associated with the proxy_username input value.
+#!                        Optional
+#! @input region_id: Region ID of an instance. You can call DescribeRegions to obtain the latest region list.
+#! @input instance_id: The specified instance ID.
+#! @input force_stop: Whether to force shutdown upon device restart.  Value range:true: force the instance to shut down
+#!                    false: the instance shuts down normallyDefault: false
```
Comment (ionutmarisca) [inline content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl]: The Default here should be on a new line

File: content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+#!                    Optional
+#! @input confirm_stop: Whether to stop an I1 ECS instance or not.  A required parameter for I1 type family instance, it
+#!                      only takes effect when the instance is of I1 type family.Valid values: true, falseDefault value:
```
Comment (ionutmarisca) [inline content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl]: The Default here should be on a new line

Comment (sushilMF) [reply content/io/cloudslang/alibaba/ecs/instances/stop_instance.sl]: Incorporated all review comments

File: content/io/cloudslang/alibaba/ecs/deploy_instance.sl [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+#!                          PasswordInherit is specified, the Password must be null. For a secure access, make sure that
+#!                          the selected image has password configured.
+#!                          Default: false
+#!                          Optional
+#! @input is_optimized: Whether it is an I/O-optimized instance or not.
+#!                      Valid values: none, optimized
+#!                      Default: optimized
+#!                      Optional
+#! @input system_disk_category: The category of the system disk.  Optional values:cloud: Basic cloud
+#!                              disk.cloud_efficiency: Ultra cloud disk.cloud_ssd: Cloud SSD.ephemeral_ssd: Ephemeral
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/deploy_instance.sl]: Please review the description for this input. Add Optional values to new line and set the default value in the input

File: content/io/cloudslang/alibaba/ecs/deploy_instance.sl [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+#!                              SSD.For phased-out instance types and non-I/O optimized instances, the default value is
+#!                              cloud.Otherwise, the default value is cloud_efficiency.
+#!                              Optional
+#! @input system_disk_size: Size of the system disk, measured in GiB. Value range: [20, 500]. The specified value must
+#!                          be equal to or greater than max{20, Imagesize}.
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/deploy_instance.sl]: Please add the default value for system_disk_size

File: content/io/cloudslang/alibaba/ecs/deploy_instance.sl [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+#!                   Optional
+#! @input client_token: It is used to guarantee the idempotence of the request. This parameter value is generated by the
+#!                      client and is guaranteed to be unique between different requests. It can contain a maximum of 64
+#!                      ASCII characters only.
+#!                      Optional
+#! @input key_pair_name: The name of the key pair.This parameter is valid only for a Linux instance. For a Windows ECS
+#!                       instance, if a value is set for parameter KeyPairName, the password still takes effect.The user
+#!                       name and password authentication method is disabled if a value is set for parameter KeyPairName
+#!                       for a Linux instance.
+#!						 Default: ''.
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/deploy_instance.sl]: Small indentation fix

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+#!                              cloud.Otherwise, the default value is cloud_efficiency.
+#!                              Optional
+#! @input system_disk_size: Size of the system disk, measured in GiB. Value range: [20, 500]. The specified value must
+#!                          be equal to or greater than max{20, Imagesize}.
+#!                          Default: max{40, ImageSize}.
+#!                          Optional
+#! @input system_disk_name: Name of the system disk.It can be [2, 128] characters in length, must begin with an English
+#!                          letter or Chinese character, and can contain digits, colons (:), underscores (_), or hyphens
+#!                          (-).The name is displayed in the ECS console.It cannot begin with http:// or https://.
+#!                          Default value: ''
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Remove value

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+#!                          Optional
+#! @input system_disk_description: Description of a system disk.It can be [2, 256] characters in length.The description
+#!                                 is displayed in the ECS console.It cannot begin with http:// or https://.
+#!                                 Default value: ''
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Remove value

File: content/io/cloudslang/alibaba/ecs/instances/create_instance.sl [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+#!                   Optional
+#! @input client_token: It is used to guarantee the idempotence of the request. This parameter value is generated by the
+#!                      client and is guaranteed to be unique between different requests. It can contain a maximum of 64
+#!                      ASCII characters only. 
+#!                      Optional
+#! @input key_pair_name: The name of the key pair. This parameter is valid only for a Linux instance. For a Windows ECS
+#!                       instance, if a value is set for parameter KeyPairName, the password still takes effect. The user
+#!                       name and password authentication method is disabled if a value is set for parameter KeyPairName
+#!                       for a Linux instance.
+#!						 Default: ''.
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/instances/create_instance.sl]: Small indentation fix

File: content/io/cloudslang/alibaba/ecs/undeploy_instance.sl [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+#! @input region_id: Region ID of an instance. You can call DescribeRegions to obtain the latest region list.
+#! @input instance_id: The ID of the instance to be terminated.
+#! @input force_stop: Whether to force shutdown upon device restart.
+#!                    Value range:true: force the instance to shut down, false: the instance shuts down normally
+#!                    Default: false
+#!                    Optional
+#! @input confirm_stop: Whether to stop an I1 ECS instance or not.  A required parameter for I1 type family instance, it
+#!                      only takes effect when the instance is of I1 type family.
+#!                      Valid values: true, false
+#!                      Default value: false
```
Comment (tethryus) [inline content/io/cloudslang/alibaba/ecs/undeploy_instance.sl]: Remove value

**Pull request merged**